fem/libceed/diffusion.h

```diff
@@ -0,0 +1,199 @@
+// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-443211. All Rights
+// reserved. See file COPYRIGHT for details.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability see http://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+
+/// A structure used to pass additional data to f_build_diff and f_apply_diff
+struct BuildContext { CeedInt dim, space_dim; CeedScalar coeff; };
+
+/// libCEED Q-function for building quadrature data for a diffusion operator
+CEED_QFUNCTION(f_build_diff_const)(void *ctx, CeedInt Q,
+                                   const CeedScalar *const *in, CeedScalar *const *out)
+{
+   BuildContext *bc = (BuildContext*)ctx;
+   // in[0] is Jacobians with shape [dim, nc=dim, Q]
+   // in[1] is quadrature weights, size (Q)
+   //
+   // At every quadrature point, compute qw/det(J).adj(J).adj(J)^T and store
+   // the symmetric part of the result.
+   const CeedScalar coeff = bc->coeff;
+   const CeedScalar *J = in[0], *qw = in[1];
+   CeedScalar *qd = out[0];
+   switch (bc->dim + 10 * bc->space_dim)
```

Comment:
libCEED doesn't currently support an unified error system for CPU and GPU. This might get added later.

Reply:
To be more precise, there use to be an error message but it has been removed in the last version of libCEED for compatibility with GPUs.

Reply:
There's nothing you can do? `std::abort` or `exit(1)` or `mfem_error()` or something? As it is this will just execute silently and the user will have no idea what is going on.

Reply:
I think the issue is that this code can be turned into a kernel on the GPU and then calling CPU functions is not possible. One solution could be to perform the check outside of this kernel and just leave a comment here that the check is in the calling scope.

Reply:
Another possible option is to use ``#if defined(__CUDA__ARCH__)`` to detect if the code will be offloaded to the GPU. If so, you can use  ``asm ("trap;")`` to halt the kernel. Or abort() on the cpu. 
There is a nice solution here: 
https://github.com/LLNL/RAJA/blob/develop/include/RAJA/config.hpp.in#L205

Reply:
You only know at runtime if the code is offloaded on GPU, I don't think that this compilation time solution would work for libCEED.

Reply:
Couldn't we use the `ctx` to get this info back?

Reply:
This is using the `ctx`